cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/Internal.hs

```diff
@@ -0,0 +1,883 @@
+foreign import ccall "blst_fp12_finalverify" c_blst_fp12_finalverify :: PTPtr -> PTPtr -> IO Bool
+
+---- Pairing
+
+foreign import ccall "blst_miller_loop" c_blst_miller_loop :: PTPtr -> Affine2Ptr -> Affine1Ptr -> IO ()
+
+---- Raw BLST error constants
+
+foreign import ccall "blst_success" c_blst_success :: CInt
+foreign import ccall "blst_error_bad_encoding" c_blst_error_bad_encoding :: CInt
+foreign import ccall "blst_error_point_not_on_curve" c_blst_error_point_not_on_curve :: CInt
+foreign import ccall "blst_error_point_not_in_group" c_blst_error_point_not_in_group :: CInt
+foreign import ccall "blst_error_aggr_type_mismatch" c_blst_error_aggr_type_mismatch :: CInt
+foreign import ccall "blst_error_verify_fail" c_blst_error_verify_fail :: CInt
+foreign import ccall "blst_error_pk_is_infinity" c_blst_error_pk_is_infinity :: CInt
+foreign import ccall "blst_error_bad_scalar" c_blst_error_bad_scalar :: CInt
+
+---- Utility functions
+
+foreign import ccall "memcmp" c_memcmp :: Ptr a -> Ptr a -> CSize -> IO CSize
+foreign import ccall "blst_bendian_from_scalar" c_blst_bendian_from_scalar :: Ptr CChar -> ScalarPtr -> IO ()
+
+data BLSTError
+  = BLST_SUCCESS
+  | BLST_BAD_ENCODING
+  | BLST_POINT_NOT_ON_CURVE
+  | BLST_POINT_NOT_IN_GROUP
+  | BLST_AGGR_TYPE_MISMATCH
+  | BLST_VERIFY_FAIL
+  | BLST_PK_IS_INFINITY
+  | BLST_BAD_SCALAR
+  | BLST_UNKNOWN_ERROR
+  deriving (Show, Eq, Ord)
+
+mkBLSTError :: CInt -> BLSTError
+mkBLSTError e
+  | e == c_blst_success
+  = BLST_SUCCESS
+  | e == c_blst_error_bad_encoding
+  = BLST_BAD_ENCODING
+  | e == c_blst_error_point_not_on_curve
+  = BLST_POINT_NOT_ON_CURVE
+  | e == c_blst_error_point_not_in_group
+  = BLST_POINT_NOT_IN_GROUP
+  | e == c_blst_error_aggr_type_mismatch
+  = BLST_AGGR_TYPE_MISMATCH
+  | e == c_blst_error_verify_fail
+  = BLST_VERIFY_FAIL
+  | e == c_blst_error_pk_is_infinity
+  = BLST_PK_IS_INFINITY
+  | e == c_blst_error_bad_scalar
+  = BLST_BAD_SCALAR
+  | otherwise
+  = BLST_UNKNOWN_ERROR
+
+-- | This class represents phantom types that can be used as BLS curves.
+class (BLS_Curve a, BLS_P a) => BLS a where
+
+instance BLS Curve1 where
+
+instance BLS Curve2 where
+
+---- Curve point operations
+
+instance BLS curve => Eq (P curve) where
+  a == b = unsafePerformIO $ do
+    withP a $ \aptr ->
+      withP b $ \bptr ->
+        c_blst_p_is_equal aptr bptr
+
+instance Eq Scalar where
+  a == b = scalarToBS a == scalarToBS b
+
+instance Eq Fr where
+  a == b = unsafePerformIO $
+    (==) <$> scalarFromFr a <*> scalarFromFr b
+
+-- | Check whether a point is on its elliptic curve.
+onCurve :: BLS_Curve curve => P curve -> Bool
+onCurve p = unsafePerformIO $ withP p c_blst_on_curve
+
+-- | Check whether a point is in the group corresponding to its elliptic curve
+inGroup :: BLS_Curve curve => P curve -> Bool
+inGroup p = unsafePerformIO $ withP p c_blst_in_g
+
+-- | Curve point addition.
+add :: (BLS curve) => P curve -> P curve -> P curve
```

Comment:
Make clear that this uses `add_or_double` not `add`

Reply:
I have now changed the name to `addOrDouble`. Better to make it explicit than to rely on documentation.